### ISSUE TITLE
RFC: MDX v2 block parsing

### ISSUE THREAD

**Issue by johno** [mdx-js/mdx#1170]
Currently, as implemented in MDX v2, the block parser begins a paragraph inside lines that _begin_ with a component. This was initially intended, but I'm not sure it makes sense in practice. It results in some unexpected behavior (IMO).

Consider the following MDX:

```mdx
# Hello, world!

<Button>Weee</Button>
```

We end up with:

```jsx
<h1>Hello, world!</h1>

<Button>
  <p>Weee</p>
</Button>
```

In v1 we result in:

```jsx
<h1>Hello, world!</h1>

<Button>Weee</Button>
```

In v1, you could work around wanting a paragraph with something like:

```mdx
# Hello, world!

<p>
  <Button>Weee</Button>
</p>
```

Another unexpected case is directly inlining your paragraph tag for the block:

```mdx
# Hello, world!

<p>
  Weee
</p>
```

Which results in:

```jsx
<h1>Hello, world!</h1>

<p>
  <p>Weee</p>
</p>
```

Why might you want to do that? It might make sense if your paragraph in MDXProvider has a styling API and you want to make a one off tweak: `<p color='red'>Weeee</p>`.

---

tl;dr I'm thinking that it likely makes the most sense to not continue block parsing _inside_ blocks which begin with JSX. It's more flexible that way.

**Comment by wooorm** [comment]
I see the problem.

But, how could one then wrap a heading, block quote, list, or whatnot inside JSX?
Where I’m going towards is: you need some way to differentiate. We had blank lines, but folks didn’t think that made sense. What would be something that *does* make sense?

**Comment by johno** [comment]
Off the top of my head perhaps that's where we continue with the Commonmark-based interleaving?

```mdx
<TomatoBox>

# I am a block!

</TomatoBox>
```

**Comment by wooorm** [comment]
Lots of the issues before were about folks expecting stuff to work w/o blank lines though 😬 E.G., https://github.com/mdx-js/mdx/issues/628#issuecomment-584444922

**Comment by johno** [comment]
> Lots of the issues before were about folks expecting stuff to work w/o blank lines though

Yeah, I think we can document the syntax and make it what we want at this point. The blank line provides a way to signal intent on how the children are and [conforms to Commonmark](https://spec.commonmark.org/0.29/#example-136). So folks getting confused in the future will primarily be a documentation problem.

Also, interestingly, what I've proposed and what the individual was confused on would still work because we begin inline parsing/tokenizing, right?

```mdx
<Note>
This *is* **markdown**!
</Note>
```

Though, granted, the heading won't be added in the following:

```mdx
<Note>
# This *is* **markdown**!
</Note>
```

🤔🤔🤔

**Comment by wooorm** [comment]
Your first example *used* to not work, doesn’t in CM/GFM, but does work in MDX@2!
Currently, in MDX@2, both your examples work: the 1st will be in a paragraph, the second in a heading.

Some other comments: https://github.com/mdx-js/mdx/issues/628#issuecomment-637011982, https://github.com/mdx-js/mdx/issues/195#issuecomment-506745859, 

One other idea, line-ending == p:

#### w/ paragraph

```
<Note>
block
</Note>
```

```
<Note>
block</Note>
```

#### w/o paragraph

```
<Note>span
</Note>
```

```
<Note>span</Note>
```

**Comment by timneutkens** [comment]
There's an interesting difference between when you do/don't want automatic paragraphs injected.

E.g.

In this case you would probably expect the markdown syntax to "just work" including adding paragraphs for the 2 lines below the title:

```mdx
<Layout>
	# Hello World!
    
    My **Blogpost

    Second Paragraph
</Layout>
```

However when you create a button like @johno was showing it makes no sense to inject an extra paragraph:

```mdx
# Hello, world!

<Button>Weee</Button>
```

Then there's also the case of "what if I want bold text in the button":

```mdx
# Hello, world!

<Button>**Weee**</Button>
```

Potentially this is a heuristic we could document, where multi-line blocks inject a paragraph for lines without headings and a single line does not.

**Comment by cwgw** [comment]
Hi all. I ended up at this thread because I've got paragraphs inside my buttons—the same behavior described by @johno in the top comment. I'm using `v2.0.0-next.7`.

I'd expected `<Button>My Button</Button` to produce a `<Button>` inside a `<p>`. I see now how it would be undesirable in many situations to wrap paragraphs around jsx, especially when using block-level components rather than buttons.

However, I struggle to imagine a situation in which I'd expect a one-liner like `<Button>My Button</Button` to produce `<Button><p>My button</p></Button>`.

> Potentially this is a heuristic we could document, where multi-line blocks inject a paragraph for lines without headings and a single line does not.

…makes a lot of sense to me.

Anyway, is there a recommended way to achieve the following?

```jsx

<p>
  <Button>My button</Button>
</p>

```

Thanks.

**Comment by wooorm** [comment]
For now, in the beta, you can use braces too I believe, as they “exit” markdown: `{<Buton>x</Buton>}` or `<Buton>{x}</Buton>`

**Comment by wooorm** [comment]
Looked at this some more today.

The solution we’re now moving to does not change the second example by John above:

```
# Hello, world!

<p>
  Weee
</p>
```

I’m also wondering what should happen here?

```
<x><y>
z
</y></x>
```

And lastly, in this case:

```
<Button>We
ee</Button>
```

**Comment by nikoladev** [comment]
> For now, in the beta, you can use braces too I believe, as they “exit” markdown: `{<Buton>x</Buton>}`

@wooorm This fix works, thanks!

---

Would something like a special prop which determines the wrapping element be a better solution? Like so:

### Default wrapper (paragraph)

Input:

```html
<Button>
  Default wrapper
</Button>
```

Output:

```html
<Button>
  <p>
    Default wrapper
  </p>
</Button>
```

### No wrapper

Input:

```html
<Button
  wrapperElement={null}
>
  No wrapper
</Button>
```

Output:

```html
<Button>
  No wrapper
</Button>
```

### Changed wrapper

Input:

```html
<Button
  wrapperElement='span'
>
  Changed wrapper
</Button>
```

Output:

```html
<Button>
  <span>
    Changed wrapper
  </span>
</Button>
```

**Comment by ChristianMurphy** [comment]
@nikoladev would you see that changing at a per-element basis?
I see that more as a global setting, which would fit better with a MDX plugin (apply once to everything, rather than adding an attribute for every element on the page).

**Comment by nikoladev** [comment]
@ChristianMurphy no, I don't see it changing for every element. So yes, a global setting sounds like a better idea :)

**Comment by ChristianMurphy** [comment]
Then it would need to be a parser change, this can be done with an MDX plugin, in version 1 or in version 2.
You can find information on plugins at https://mdxjs.com/guides/writing-a-plugin and https://unifiedjs.com/learn
If you have follow up questions on plugins, starting a thread at https://github.com/mdx-js/mdx/discussions would be a good place to discuss.

**Comment by nikoladev** [comment]
@ChristianMurphy thanks, I'll try it out!

Am I correct in assuming that [this plugin](https://github.com/remarkjs/remark-unwrap-images) already does what I want (but then only for images)? Because then I would only need to modify that code.

**Comment by nikoladev** [comment]
> Am I correct in assuming that [this plugin](https://github.com/remarkjs/remark-unwrap-images) already does what I want (but then only for images)? Because then I would only need to modify that code.

To answer my own question: yes. For anybody else that wants to do remove all wrapping paragraphs from their imported components:

- Clone [remark-unwrap-images](https://github.com/remarkjs/remark-unwrap-images)
- Change `index.js` to the code below
- Load your plugin

```js
var visit = require('unist-util-visit')

module.exports = unwrapEverything

var splice = [].splice

function unwrapEverything () {
  return transform
}

function transform (tree) {
  visit(tree, 'paragraph', onparagraph)
}

function onparagraph (node, index, parent) {
  splice.apply(parent.children, [ index, 1 ].concat(node.children))
  return [ visit.SKIP, index ]
}
```

@ChristianMurphy thanks for the help :)

**Comment by michrome** [comment]
For some of the above examples, would it be helpful to think of the component as inline tags?. For example this input:

<pre>This is a long sentence that is
&lt;span&gt;broken&lt;/span&gt; over two lines.</pre>

should, according to the [markdown dingus](https://daringfireball.net/projects/markdown/dingus),  render to:

<pre>&lt;p&gt;This is a long sentence that is
&lt;span&gt;broken&lt;/span&gt; over two lines.&lt;/p&gt;</pre>

That will render in a browser as one line like:

> This is a long sentence that is broken over two lines. ✅

----

Unfortunately, the [MDX Playground]() is taking the same input:

<pre>This is a long sentence that is
&lt;span&gt;broken&lt;/span&gt; over two lines.</pre>

and rendering it to:

<pre>&lt;p&gt;This is a long sentence that is&lt;/p&gt;
&lt;span&gt;broken&lt;/span&gt; over two lines.</pre>

That will render in the browser, unexpectedly, as two lines like:

> This is a long sentence that is
> broken over two lines. ❌

----

Given the above, I'd say the block parser behaves unexpectedly.

**Comment by michrome** [comment]
> For now, in the beta, you can use braces too I believe, as they “exit” markdown: `{<Buton>x</Buton>}` or `<Buton>{x}</Buton>`

I tried:

<pre>This is a long sentence that is
{&lt;span&gt;broken&lt;/span&gt;} over two lines.</pre>

It is removing the line break but the braces are showing.

> This is a long sentence that is {broken} over two lines. ❌

**Comment by michrome** [comment]
Third post in a row from me as I thought this might be a good candidate for a test case.

Given the component `components/html-tag.js`:

```javascript
export default function HtmlTag({ t }) {
  return <code>&lt;{t}&gt;</code>;
}
```

One would expect the following `pages/example.mdx`, where no line is longer than 80 chars, to render in a single paragraph:

```md
import HtmlTag from "../components/html-tag";

When writing HTML, keeping the document valid is a good first step towards
accessibility. Compared to XHTML, HTML5 validation is surprisingly lenient,
requiring only a <HtmlTag t="DOCTYPE" /> and a <HtmlTag t="title" /> element.
The document below is valid even though it doesn’t contain an
<HtmlTag t="html" />, <HtmlTag t="head" /> or <HtmlTag t="body" /> element,
because they’re all considered optional. Also optional is the closing
<HtmlTag t="/p" /> tag.
```

---

HTML string:

```html
<p>When writing HTML, keeping the document valid is a good first step towards accessibility. Compared to XHTML, HTML5 validation is surprisingly lenient, requiring only a <code>&lt;DOCTYPE&gt;</code> and a <code>&lt;title&gt;</code> element. The document below is valid even though it doesn’t contain an <code>&lt;html&gt;</code>, <code>&lt;head&gt;</code> or <code>&lt;body&gt;</code> element, because they’re all considered optional. Also optional is the closing <code>&lt;/p&gt;</code> tag.</p>
```

Rendered in browser:

<p>When writing HTML, keeping the document valid is a good first step towards accessibility. Compared to XHTML, HTML5 validation is surprisingly lenient, requiring only a <code>&lt;DOCTYPE&gt;</code> and a <code>&lt;title&gt;</code> element. The document below is valid even though it doesn’t contain an <code>&lt;html&gt;</code>, <code>&lt;head&gt;</code> or <code>&lt;body&gt;</code> element, because they’re all considered optional. Also optional is the closing <code>&lt;/p&gt;</code> tag.</p>

**Comment by wooorm** [comment]
@michrome thanks for your feedback! I’ve hidden them though, because it seems your feedback is specific to MDX@1. We are aware MDX@1 is significantly limited in certain cases, hence this issue about MDX@2. You can try it out on the `next` branch; do let us know what you think of it!

**Comment by michrome** [comment]
Hey @wooorm is [https://mdxjs.com/playground](https://mdxjs.com/playground) running the `next` branch? Sorry if I have misunderstood.

<img width="970" alt="Screenshot 2020-11-18 at 08 33 45" src="https://user-images.githubusercontent.com/596770/99505520-d6b60700-2978-11eb-9ecd-b3c8b2d655b6.png">

**Comment by wooorm** [comment]
It is not. Our docs still point to MDX@1 but our work and this thread are focussed on MDX@2

**Comment by michrome** [comment]
Thanks for confirming. I just tried it with `next.8` and it rendered exactly what I wanted. 😍

**Comment by wooorm** [comment]
Hi all! I’m going to close this as it landed on the `main`, the (now default) branch we’re using to gear up to the MDX@2 release.

The reason is so that it’s more clear which issues still need to be solved already and don’t require further work.

Expect another `next` release soonish, which might include it, or maybe it’s already released as a beta!

For more info, see <https://github.com/mdx-js/mdx/issues/1041>.
